src/blockrelay/mempool_sync.h

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2018-2019 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_MEMPOOL_SYNC_H
+#define BITCOIN_MEMPOOL_SYNC_H
+
+#include "blockrelay/blockrelay_common.h"
+#include "blockrelay/graphene.h"
+#include "consensus/consensus.h"
+
+#include <chrono>
+
+const uint64_t DEFAULT_MEMPOOL_SYNC_MIN_VERSION_SUPPORTED = 0;
+const uint64_t DEFAULT_MEMPOOL_SYNC_MAX_VERSION_SUPPORTED = 0;
+// arbitrary entropy passed to CGrapheneSet an used for IBLT
+const uint32_t IBLT_ENTROPY = 13;
+// any value greater than 2 will use SipHash
+const uint64_t SHORT_ID_VERSION = 2;
+// frequency of synchronization (per peer) in microseconds
+const int64_t MEMPOOLSYNC_FREQ_US = 30 * 1e6;
+const int64_t MEMPOOLSYNC_FREQ_GRACE_US = 5 * 1e6;
+// Use CVariableFastFilter if true, otherwise use CBloomFilter
+const bool COMPUTE_OPTIMIZED = true;
+
+class CMempoolSyncState
+{
+public:
+    std::chrono::time_point<std::chrono::high_resolution_clock> lastUpdated;
+    uint64_t shorttxidk0;
+    uint64_t shorttxidk1;
+    bool completed;
+
+public:
+    CMempoolSyncState(std::chrono::time_point<std::chrono::high_resolution_clock> _lastUpdated,
+        uint64_t _shorttxidk0,
+        uint64_t _shorttxidk1,
+        bool _completed)
+        : lastUpdated(_lastUpdated), shorttxidk0(_shorttxidk0), shorttxidk1(_shorttxidk1), completed(_completed)
+    {
+    }
+    CMempoolSyncState()
+        : lastUpdated(std::chrono::high_resolution_clock::now()), shorttxidk0(0), shorttxidk1(0), completed(false)
+    {
+    }
+};
+
+extern std::map<CNode *, CMempoolSyncState> mempoolSyncRequested;
```

Comment:
If you use a pointer as the key, and then dereference this pointer, you need to make sure that the CNode* pointer is not deleted.  CNode* pointers can be deleted whenever a node disconnects.  To do so, you'd need to call CNode::AddRef(), and hold that reference over the entire time CNode* pointers are stored in your map.  But if you do this, you'll prevent the cleanup of CNodes, so you need to implement code that periodically searches your maps for any disconnected CNodes, deletes the entry from your map and releases the ref.

Another possibility (which is more common in your situation) is to use the node id.  This id is a non-repeating number.  You use CNodeRef CConnMgr::FindNodeFromId(NodeId id) to get the CNode (and automatically AddRef(), btw) and use CNode::GetId() to get the id.  FindNodeFromId will return a null reference if the node has been deleted.  This is better than using CNode* since you are not holding up the node's cleanup when it disconnects.  You'll still need to periodically go thru your maps and clean out any entries whose nodeId doesn't match a live node in case a node disconnects in the middle of a mempool sync op.  Otherwise a connect, start sync, disconnect attack could fill your maps with entries.  However, this cleanup could happen rarely.

Reply:
Done. I went with NodeId and also added a cleanup method that is called from net.cpp.